### ISSUE TITLE
some tests, sageattention is slowest one

### ISSUE THREAD

**Issue by ptits** [kijai/ComfyUI-MochiWrapper#36]
sdpa       [01:38<00:00,  1.97s/it]
flash_attn [01:35<00:00,  1.91s/it]
sage_attn  [01:43<00:00,  2.06s/it]

H100nvcc -V
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2024 NVIDIA Corporation
Built on Wed_Aug_14_10:10:22_PDT_2024
Cuda compilation tools, release 12.6, V12.6.68
Build cuda_12.6.r12.6/compiler.34714021_0

pip list 
Package                   Version                  Editable project location
------------------------- ------------------------ -------------------------
absl-py                   2.1.0
accelerate                1.0.1
addict                    2.4.0
aiofiles                  22.1.0
aiohappyeyeballs          2.4.3
aiohttp                   3.10.10
aiosignal                 1.3.1
aiosqlite                 0.20.0
albucore                  0.0.17
albumentations            1.4.18
altair                    5.4.1
annotated-types           0.7.0
antlr4-python3-runtime    4.9.3
anyio                     4.3.0
archspec                  0.2.3
argon2-cffi               23.1.0
argon2-cffi-bindings      21.2.0
argostranslate            1.9.6
arrow                     1.3.0
asttokens                 2.4.1
async-lru                 2.0.4
attrs                     23.2.0
Babel                     2.14.0
beautifulsoup4            4.12.3
bleach                    6.1.0
bokeh                     3.3.4
boltons                   24.0.0
Brotli                    1.1.0
cached-property           1.5.2
cachetools                5.3.3
certifi                   2024.2.2
cffi                      1.16.0
chardet                   5.2.0
charset-normalizer        3.3.2
click                     8.1.7
cloudpickle               2.2.1
color-matcher             0.5.0
colorama                  0.4.6
colorlog                  6.8.2
comm                      0.2.2
conda                     24.5.0
conda-libmamba-solver     24.1.0
conda-package-handling    2.2.0
conda_package_streaming   0.9.0
contourpy                 1.2.1
cryptography              43.0.3
cssselect2                0.7.0
ctranslate2               4.4.0
cycler                    0.12.1
Cython                    3.0.11
ddt                       1.7.2
debugpy                   1.8.1
decorator                 5.1.1
defusedxml                0.7.1
Deprecated                1.2.14
dill                      0.3.8
distro                    1.9.0
docstring_parser          0.16
docutils                  0.21.2
easydict                  1.13
einops                    0.8.0
embreex                   2.17.7.post5
entrypoints               0.4
eval_type_backport        0.2.0
exceptiongroup            1.2.0
executing                 2.0.1
fastapi                   0.115.3
fastjsonschema            2.19.1
ffmpy                     0.4.0
filelock                  3.13.1
flash_attn                2.6.3
flatbuffers               24.3.25
fonttools                 4.52.1
fqdn                      1.5.1
frozendict                2.4.4
frozenlist                1.4.1
fsspec                    2024.2.0
ftfy                      6.3.0
fvcore                    0.1.5.post20221221
gitdb                     4.0.11
GitPython                 3.1.43
google-api-core           2.19.0
google-auth               2.29.0
google-cloud-core         2.4.1
google-cloud-storage      2.16.0
google-crc32c             1.5.0
google-resumable-media    2.7.0
googleapis-common-protos  1.63.0
gradio                    3.36.1
gradio_client             1.4.2
h11                       0.14.0
h2                        4.1.0
hpack                     4.0.0
httpcore                  1.0.5
httpx                     0.27.0
huggingface-hub           0.26.1
hyperframe                6.0.1
idna                      3.7
imagecodecs               2024.1.1
imageio                   2.34.1
imageio-ffmpeg            0.5.1
importlib_metadata        7.1.0
importlib_resources       6.4.0
insightface               0.7.3
iopath                    0.1.10
ipykernel                 6.29.3
ipympl                    0.9.4
ipython                   8.24.0
ipython_genutils          0.2.0
ipywidgets                8.1.2
isoduration               20.11.0
jax                       0.4.34
jaxlib                    0.4.34
jedi                      0.19.1
Jinja2                    3.1.4
joblib                    1.4.2
json5                     0.9.25
jsonpatch                 1.33
jsonpointer               2.4
jsonschema                4.22.0
jsonschema-specifications 2023.12.1
jupyter_client            7.4.9
jupyter_core              5.7.2
jupyter-events            0.10.0
jupyter-lsp               2.2.5
jupyter-resource-usage    0.7.2
jupyter_server            2.14.0
jupyter_server_fileid     0.9.2
jupyter-server-mathjax    0.2.6
jupyter_server_terminals  0.5.3
jupyter_server_ydoc       0.8.0
jupyter-ydoc              0.2.5
jupyterlab                3.6.7
jupyterlab-execute-time   2.3.1
jupyterlab_git            0.50.1
jupyterlab_pygments       0.3.0
jupyterlab_server         2.27.2
jupyterlab-system-monitor 0.8.0
jupyterlab-topbar         0.6.1
jupyterlab_widgets        3.0.10
kfp                       2.7.0
kfp-pipeline-spec         0.3.0
kfp-server-api            2.0.5
kiwisolver                1.4.5
kornia                    0.7.3
kornia_rs                 0.1.5
kubernetes                26.1.0
lazy_loader               0.4
libmambapy                1.5.8
linkify-it-py             2.0.3
lxml                      5.3.0
mamba                     1.5.8
manifold3d                2.5.1
mapbox_earcut             1.0.2
markdown-it-py            2.2.0
MarkupSafe                2.1.5
matplotlib                3.8.4
matplotlib-inline         0.1.7
matrix-client             0.4.0
mdit-py-plugins           0.3.3
mdurl                     0.1.2
mediapipe                 0.10.15
menuinst                  2.1.0
mistune                   3.0.2
ml_dtypes                 0.5.0
mochi                     0.1.0                    /home/jovyan/models
mpmath                    1.3.0
msgpack                   1.1.0
mss                       9.0.2
multidict                 6.1.0
munkres                   1.1.4
narwhals                  1.10.0
nbclassic                 1.1.0
nbclient                  0.10.0
nbconvert                 7.16.4
nbdime                    4.0.1
nbformat                  5.10.4
nest_asyncio              1.6.0
networkx                  3.4.2
notebook                  6.5.7
notebook_shim             0.2.4
numpy                     1.26.4
nvidia-cublas-cu12        12.4.5.8
nvidia-cuda-cupti-cu12    12.4.127
nvidia-cuda-nvrtc-cu12    12.4.127
nvidia-cuda-runtime-cu12  12.4.127
nvidia-cudnn-cu12         9.1.0.70
nvidia-cufft-cu12         11.2.1.3
nvidia-curand-cu12        10.3.5.147
nvidia-cusolver-cu12      11.6.1.9
nvidia-cusparse-cu12      12.3.1.170
nvidia-cusparselt-cu12    0.6.2
nvidia-ml-py3             7.352.0
nvidia-nccl-cu12          2.21.5
nvidia-nvjitlink-cu12     12.4.127
nvidia-nvtx-cu12          12.4.127
oauthlib                  3.2.2
omegaconf                 2.3.0
onnx                      1.17.0
opencv-contrib-python     4.10.0.84
opencv-python             4.10.0.84
opencv-python-headless    4.10.0.84
opt_einsum                3.4.0
orjson                    3.10.10
overrides                 7.7.0
packaging                 24.0
pandas                    2.2.3
pandocfilters             1.5.0
parso                     0.8.4
patsy                     0.5.6
pexpect                   4.9.0
pickleshare               0.7.5
piexif                    1.1.3
pillow                    11.0.0
pip                       24.0
pkgutil_resolve_name      1.3.10
platformdirs              4.2.2
pluggy                    1.5.0
ply                       3.11
portalocker               2.10.1
prettytable               3.11.0
prometheus_client         0.20.0
prompt-toolkit            3.0.42
propcache                 0.2.0
proto-plus                1.23.0
protobuf                  4.25.3
psutil                    5.9.8
ptyprocess                0.7.0
pure-eval                 0.2.2
pyasn1                    0.6.0
pyasn1_modules            0.4.0
pycollada                 0.8
pycosat                   0.6.6
pycparser                 2.22
pydantic                  2.9.2
pydantic_core             2.23.4
pydub                     0.25.1
PyGithub                  2.4.0
Pygments                  2.18.0
PyJWT                     2.9.0
PyNaCl                    1.5.0
pynvml                    11.5.3
pyparsing                 3.1.2
PyQt5                     5.15.9
PyQt5-sip                 12.12.2
PySocks                   1.7.1
python-dateutil           2.9.0
python-json-logger        2.0.7
python-multipart          0.0.12
pytorch-triton            3.1.0+cf34004b8a
pytz                      2024.1
PyWavelets                1.4.1
PyYAML                    6.0.2
pyzmq                     26.0.3
ray                       2.37.0
referencing               0.35.1
regex                     2024.9.11
reportlab                 4.2.5
requests                  2.32.2
requests-oauthlib         2.0.0
requests-toolbelt         0.10.1
rfc3339-validator         0.1.4
rfc3986-validator         0.1.1
rich                      13.9.2
rpds-py                   0.18.1
rsa                       4.9
Rtree                     1.3.0
ruamel.yaml               0.18.6
ruamel.yaml.clib          0.2.8
sacremoses                0.0.53
safetensors               0.4.5
sageattention             1.0.3
scikit-image              0.22.0
scikit-learn              1.3.2
scipy                     1.11.3
seaborn                   0.13.2
semantic-version          2.10.0
Send2Trash                1.8.3
sentencepiece             0.2.0
setuptools                75.2.0
shapely                   2.0.6
shellingham               1.5.4
sip                       6.7.12
six                       1.16.0
smmap                     5.0.1
sniffio                   1.3.1
sounddevice               0.5.1
soundfile                 0.12.1
soupsieve                 2.5
spandrel                  0.4.0
stack-data                0.6.2
stanza                    1.1.1
starlette                 0.41.0
statsmodels               0.14.2
svg.path                  6.3
svglib                    1.5.1
sympy                     1.13.1
tabulate                  0.9.0
termcolor                 2.5.0
terminado                 0.18.1
threadpoolctl             3.5.0
tifffile                  2024.5.22
tinycss2                  1.3.0
tokenizers                0.20.1
toml                      0.10.2
tomli                     2.0.1
torch                     2.6.0.dev20241025+cu124
torchaudio                2.5.0.dev20241025+cu124
torchsde                  0.2.6
torchvision               0.20.0.dev20241025+cu124
tornado                   6.4
tqdm                      4.66.4
traitlets                 5.14.3
trampoline                0.1.2
transformers              4.45.2
trimesh                   4.5.0
triton                    3.1.0
truststore                0.8.0
typer                     0.12.5
types-python-dateutil     2.9.0.20240316
typing_extensions         4.11.0
typing-utils              0.1.0
tzdata                    2024.1
uc-micro-py               1.0.3
uri-template              1.3.0
urllib3                   1.26.18
uv                        0.4.25
uvicorn                   0.32.0
vhacdx                    0.0.8.post1
wcwidth                   0.2.13
webcolors                 1.13
webencodings              0.5.1
websocket-client          1.8.0
websockets                12.0
wheel                     0.43.0
widgetsnbextension        4.0.10
wrapt                     1.16.0
xatlas                    0.0.9
xgboost                   1.7.6
xxhash                    3.5.0
xyzservices               2024.4.0
y-py                      0.6.2
yacs                      0.1.8
yapf                      0.40.2
yarl                      1.15.5
ypy-websocket             0.8.4
zipp                      3.17.0
zstandard                 0.19.0

**Comment by ptits** [comment]
https://github.com/user-attachments/assets/3d85d16e-eebc-4449-a13f-7e936abd11bb

Here is three merged clips spda, flash_attn, sage_attn - looks like flasg and sage are the same

**Comment by ptits** [comment]
![shot_241028_001456](https://github.com/user-attachments/assets/b526d351-f8d8-4422-ba5f-a969acad33f4)

**Comment by kijai** [comment]
Flash attention and the new "Fused Flash Attention" which is used with the cudnn sdpa kernel seem to perform better on H100 indeed.

**Comment by jpgallegoar** [comment]
Did you enable sage in "Mochi Model Loader" node -> attention mode?
